### PR TITLE
[Merged by Bors] - feat: additions to ListM API

### PR DESCRIPTION
**Pull request by kim-em**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/ListM/Basic.lean [hunk 1]
```diff
@@ -103,15 +126,33 @@ unsafe def force {α} : ListM m α → m (List α)
...
+    (· :: ·) x <$> takeAsList xs n
+#align tactic.mllist.take ListM.takeAsList
+
+/-- Take the first `n` elements. -/
+unsafe def take : ListM m α → Nat → ListM m α
+  | _, 0 => empty
+  | nil, _ => empty
+  | cons l, n+1 => cons do
+    let (a, l) ← l
+    let some a ← pure a | return (none, l.take (n+1))
```
Comment (gebner) [inline Mathlib/Data/ListM/Basic.lean:145]:
```suggestion
    let some a := a | return (none, l.take (n+1))
```

File: Mathlib/Data/ListM/Basic.lean [hunk 2]
```diff
@@ -103,15 +126,33 @@ unsafe def force {α} : ListM m α → m (List α)
...
+    return (a, l.take n)
+
+/-- Drop the first `n` elements. -/
+unsafe def drop : ListM m α → Nat → ListM m α
+  | e, 0 => e
+  | nil, _ => nil
+  | cons l, n+1 => cons do
+    let (a, l) ← l
+    let some _ ← pure a | return (none, drop l (n+1))
```
Comment (gebner) [inline Mathlib/Data/ListM/Basic.lean:154]:
```suggestion
    let some _ := a | return (none, drop l (n+1))
```

File: Mathlib/Data/ListM/Basic.lean [hunk 3]
```diff
@@ -103,15 +126,33 @@ unsafe def force {α} : ListM m α → m (List α)
...
+#align tactic.mllist.take ListM.takeAsList
+
+/-- Take the first `n` elements. -/
+unsafe def take : ListM m α → Nat → ListM m α
+  | _, 0 => empty
+  | nil, _ => empty
+  | cons l, n+1 => cons do
+    let (a, l) ← l
+    let some a ← pure a | return (none, l.take (n+1))
+    return (a, l.take n)
```
Comment (gebner) [inline Mathlib/Data/ListM/Basic.lean:146]: These are probably written more easily using `uncons`:
```suggestion
  | _, 0 => empty
  | xs, n+1 => cons do
    match ← uncons xs with
    | some (x, xs) => return (x, xs.take n)
    | none => return (none, empty)
```

File: Mathlib/Data/ListM/Basic.lean [hunk 4]
```diff
@@ -248,21 +311,79 @@ unsafe def bind {α β : Type u} : ListM m α → (α → ListM m β) → ListM
...
+#align tactic.mllist.monad_lift ListM.monadLift
+
+/-- Lift the monad of a lazy list. -/
+unsafe def liftM [Monad n] [MonadLift m n] (L : ListM m α) : ListM n α :=
+squash do match ← (uncons L : m _) with
+  | none => pure empty
+  | some (a, L') => pure <| cons do pure (a, L'.liftM)
+
+/-- Given a lazy list in a state monad, run it on some initial state, recording the states. -/
+unsafe def run {σ α : Type u} (L : ListM (StateT.{u} σ m) α) (s : σ) : ListM m (α × σ) :=
```
Comment (gebner) [inline Mathlib/Data/ListM/Basic.lean:325]:
```suggestion
unsafe def runState {σ α : Type u} (L : ListM (StateT.{u} σ m) α) (s : σ) : ListM m (α × σ) :=
```
I imagine run-like methods would be useful for other monad transformers as well.

File: Mathlib/Data/ListM/Basic.lean [hunk 5]
```diff
@@ -248,21 +311,79 @@ unsafe def bind {α β : Type u} : ListM m α → (α → ListM m β) → ListM
...
+squash do match ← StateT.run (uncons L) s with
+  | (none, _) => pure empty
+  | (some (a, L'), s') => pure <| cons do pure (some (a, s'), L'.run s')
+
+/-- Given a lazy list in a state monad, run it on some initial state. -/
+unsafe def run' {σ α : Type u} (L : ListM (StateT.{u} σ m) α) (s : σ) : ListM m α :=
```
Comment (gebner) [inline Mathlib/Data/ListM/Basic.lean:331]:
```suggestion
unsafe def runState' {σ α : Type u} (L : ListM (StateT.{u} σ m) α) (s : σ) : ListM m α :=
```

File: Mathlib/Data/ListM/Basic.lean [hunk 6]
```diff
@@ -36,15 +36,35 @@ namespace ListM
 
 variable {α β : Type u} {m : Type u → Type u}
 
+unsafe instance : EmptyCollection (ListM m α) := ⟨nil⟩
+
+/-- The implementation of `ForIn`, which enables `for a in L do ...` notation. -/
+@[specialize] protected unsafe def forIn [Monad m] [Monad n] [MonadLiftT m n]
+    (as : ListM m α) (init : δ) (f : α → δ → n (ForInStep δ)) : n δ :=
+match as with
```
Comment (gebner) [inline Mathlib/Data/ListM/Basic.lean:44]:
```suggestion
  match as with
```

We indent function bodies in Lean 4.

File: Mathlib/Data/ListM/Basic.lean [hunk 7]
```diff
@@ -180,6 +221,22 @@ unsafe def filterMapM [Alternative m] {α β : Type u} (f : α → m β) : ListM
       (f a >>= fun b => return (some b, filterMapM f r)) <|> return (none, filterMapM f r)
 #align tactic.mllist.mfilter_map ListM.filterMapM
 
+/-- Take the initial segment of the lazy list, until the function `f` first fails. -/
+unsafe def takeWhileM [Alternative m] (f : α → m β) : ListM m α → ListM m α
```
Comment (gebner) [inline Mathlib/Data/ListM/Basic.lean:225]: I feel like checking for success is a bit of a Lean 3-ism, an API like `takeWhileM (f : α → m (ULift Bool)) ...` is more natural now and also consistent with `List.filterM`.

Comment (semorrison) [reply Mathlib/Data/ListM/Basic.lean:225]: Okay. I liked that Lean 3-ism, but I agree it is not necessarily the obvious semantics!

File: Mathlib/Data/ListM/Basic.lean (hunk 3 above)
Comment (semorrison) [reply Mathlib/Data/ListM/Basic.lean]: Oh, yes. I've rewritten all the functions that benefit from this.

File: Mathlib/Data/ListM/Basic.lean (hunk 7 above)
Comment (semorrison) [reply Mathlib/Data/ListM/Basic.lean:225]: Hmm, I really like the `filterMapM` for being able to discard cases where the monadic function fails. I feel like every time I call this it is going to be with a failing tactic that with this change I need to `try ... catch ...`.

What about a primed version of these, that assume `[Alternative m]` and have the old behaviour?

File: Mathlib/Data/ListM/Basic.lean (hunk 6 above)
Comment (semorrison) [reply Mathlib/Data/ListM/Basic.lean]: Ooooops. Fixed.

**Comment by semorrison** [comment]
Thanks for the review, @gebner. I'll go either way on the `filterMapM` question, but would like you to say "yes, really" before I fold on this one. :-)

File: Mathlib/Data/ListM/Basic.lean (hunk 7 above)
Comment (gebner) [reply Mathlib/Data/ListM/Basic.lean:199]: > I feel like every time I call this it is going to be with a failing tactic that with this change I need to try ... catch ....

We have a `succeeds` function for that, and `xs.takeWhileM (succeeds do checkType ·)` looks pretty nice and explicit.

But I don't want to block this PR on that change.  Most functions on `ListM` have this implicit succeeds-API.  I know you that you need the PR for more interesting code. :smile:

Comment (gebner) [reply Mathlib/Data/ListM/Basic.lean:199]: Oh and for `filterMapM` there's `try?` to convert failure to Option.

Comment (semorrison) [reply Mathlib/Data/ListM/Basic.lean]: Okay, I'm convinced, and have changed the API.

**Comment by semorrison** [comment]
bors merge

File: Mathlib/Data/ListM.lean
Comment (eric-wieser) [inline Mathlib/Data/ListM.lean:1]: Can you split this PR in two? Moving and changing the file at the same time makes it hard to follow the diff.

Comment (eric-wieser) [reply Mathlib/Data/ListM.lean:1]: bors r-

(since this isn't in the active batch right now anyway)

**Comment by eric-wieser** [comment]
bors merge
